File: Templates/NumberTheory/SieveOfEuler.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
"""
 @file Templates/NumberTheory/SieveOfEuler.py
 @version 0.3
 @author SHawnHardy
 @date 2020-04-26
 @copyright MIT License
 @last_used ProjectEuler/0012.py
"""

import sys

sys.setrecursionlimit(1000000)


def sieve_of_euler(limit):
    prime = []
    is_prime = [True for _ in range(limit)]
    is_prime[0] = is_prime[1] = False
    prime_factor = [_ for _ in range(limit)]

    for x in range(2, limit):
        if is_prime[x]:
            prime.append(x)

        for p in prime:
            if p * x >= limit:
                break
            is_prime[p * x] = False
            prime_factor[p * x] = p
            if x % p == 0:
                break
    return prime, is_prime, prime_factor

File: Templates/NumberTheory/SieveOfEuler.cpp
/*
 * @file Templates/NumberTheory/SieveOfEuler.cpp
 * @version 0.1
 * @author SHawnHardy
 * @date 2020-04-26
 * @copyright MIT License
 * @last_used ProjectEuler/0003.cpp
 */

#include "algorithm"
#include "cmath"
#include "complex"
#include "cstdio"
#include "cstdlib"
#include "cstring"
#include "iostream"
#include "map"
#include "queue"
#include "set"
#include "stack"
#include "string"
#include "vector"

#define pb push_back
#define fi first
#define se second
#define dbg(...)                                                               \
    cerr << "[" << #__VA_ARGS__ ":" << (__VA_ARGS__) << "]" << endl;

using namespace std;
typedef long long int ll;
typedef pair<int, int> pii;

const int MAXN = 200 + 10;
const int MOD = 1E9 + 7;
const int INF = 0x3f3f3f3f;

struct SieveOfEuler {
    vector<int> prime;
    bool *is_prime;
    int *prime_factor;

    SieveOfEuler(int limit) {
        prime.clear();
        is_prime = new bool[limit]{};
        memset(is_prime, -1, limit);
        is_prime[0] = is_prime[1] = false;
        prime_factor = new int[limit]{};

        for (int i = 2; i < limit; i++) {
            if (is_prime[i]) {
                prime.push_back(i);
                prime_factor[i] = i;
            }

            for (auto j = prime.begin(); j != prime.end() && (*j) * i < limit;
                 j++) {
                is_prime[(*j) * i] = false;
                prime_factor[(*j) * i] = *j;
                if (i % (*j) == 0)
                    break;
            }
        }
    }

    ~SieveOfEuler() {
        delete is_prime;
        delete prime_factor;
    }
};